service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/SdsFilter.java

```diff
@@ -152,14 +135,58 @@ private String extractSspTraceId(HttpHeaders httpHeaders) {
         throw new SdsException("Missing Ssp-TraceID Header for X-Correlation-Id for SDS Request");
     }
 
+    private Mono<SdsClient.SdsResponseData> performRequestAccordingToInteractionId(String interactionId,
+        String organisation, String sspTraceId, ServerWebExchange exchange) {
+        if (sdsRequestFunctions.containsKey(interactionId)) {
+            LoggingUtil.info(LOGGER, exchange, "Performing request with organisation \"{}\" and NHS service endpoint id \"{}\"",
+                organisation, interactionId);
+            return sdsRequestFunctions.get(interactionId)
+                .apply(organisation, sspTraceId, exchange);
+        }
+        throw new IllegalArgumentException(String.format("Not recognised InteractionId %s", interactionId));
+    }
+
     private Optional<URI> prepareLookupUri(String address, ServerHttpRequest serverHttpRequest) {
```

Comment:
I've proposed something like this to find the last occurrence of either 'Binary' or 'Patient' in the path. Disclaimer: haven't tried it, no idea if it actually works.

```
private Optional<URI> prepareLookupUri(String serviceRootUrl, ServerHttpRequest originalRequest) {
    var originalRequestPath = originalRequest.getPath();
    var originalRequestPathValues = originalRequestPath.elements().stream()
        .map(PathContainer.Element::value)
        .collect(Collectors.toList());
    int indexOfPatientInFhirPath = originalRequestPathValues.lastIndexOf("Patient");
    int indexOfBinaryInFhirPath = originalRequestPathValues.lastIndexOf("Binary");
    int indexOfStartOfFhirPath = Math.max(indexOfPatientInFhirPath, indexOfBinaryInFhirPath);
    if(indexOfStartOfFhirPath < 0) {
        throw new RuntimeException("Choose a better class and error message");
    }
    String fhirRequestPathPart = originalRequest.getPath().subPath(indexOfStartOfFhirPath)
        .toString()
        .substring(1);
    String uriWithoutQueryParameters = serviceRootUrl + fhirRequestPathPart;
    URI constructedUri = UriComponentsBuilder.fromUriString(uriWithoutQueryParameters)
        .queryParams(originalRequest.getQueryParams())
        .build()
        .toUri();
    return Optional.of(constructedUri);
}
```